libraries/tock-register-interface/README.md

```diff
@@ -3,130 +3,107 @@
-    cr: ReadWrite<u8, Control::Register>,
-
-    // Status register: read-only
-    s: ReadOnly<u8, Status::Register>
-
-    // Registers can be bytes, halfwords, or words:
-    // Note that the second type parameter can be omitted, meaning that there
-    // are no bitfields defined for these registers.
-    byte0: ReadWrite<u8>,
-    byte1: ReadWrite<u8>,
-    short: ReadWrite<u16>,
+The `peripheral!` macro tests that you specified the fields (registers and
+padding) in increasing order, with no gaps.
 
-    // The padding length was automatically computed as 0x008 - 0x006.
-    _reserved: [u8; 2],
-    word: ReadWrite<u32>,
+The macro generates a trait representing the peripheral with one method for each
+register:
 
-    // Arrays are expanded as-is, like any other type.
-    array: [ReadWrite<u32>; 4],
-
-    // Etc.
+```rust
+trait Registers {
+    // Type aliases (cr, so, byte0) omitted for brevity.
+    fn cr(&self) -> Self::cr { ... }
+    fn s(&self) -> Self::s { ... }
+    fn byte0(&self) -> Self::byte0 { ... }
+    fn byte1(&self) -> Self::byte1 { ... }
+    fn short(&self) -> Self::short { ... }
+    fn array(&self) -> Self::array { ... }
 }
 ```
 
-This crate will generate additional, compile time (`const`) assertions
-to validate various invariants of the register structs, such as
-
-- proper start offset of padding fields,
-- proper start and end offsets of actual fields,
-- invalid alignment of field types,
-- the `@END` marker matching the size of the struct.
+The types returned by the `Registers` methods implement traits defined in the
+`tock_registers::register_traits` module. `peripheral!` also generates an
+implementation of the trait for use on the real chip:
 
-For more information on the generated assertions, check out the [`test_fields!`
-macro documentation](https://docs.tockos.org/tock_registers/macro.test_fields.html).
+```rust
+struct RealRegisters { ... }
 
-By default, the visibility of the generated structs and fields is private. You
-can make them public using the `pub` keyword, just before the struct name or the
-field identifier.
+impl Registers for RealRegisters { ... }
+```
 
-For example, the following call to the macro:
+Drivers can use the trait to work with both the real hardware and with fake
+hardware for unit testing:
 
 ```rust
-register_structs! {
-    pub Registers {
-        (0x000 => foo: ReadOnly<u32>),
-        (0x004 => pub bar: ReadOnly<u32>),
-        (0x008 => @END),
+struct Driver<R: Registers> {
+    registers: R,
+}
+
+// Real systems will use Driver<RealRegisters>
+
+// Unit tests will provide:
+struct Fake { ... }
+impl Registers for Fake {
+    fn cr(&self) => Self::cr {
+        tock_registers::FakeRegister::with_data(self)
+            .on_read(|this| ...)
+            .on_write(|this| ...)
     }
 }
+// then use Driver<Fake>.
 ```
 
-will generate the following struct.
+It also generates a `macro_rules!` macro with the same name as the trait:
```

Comment:
What's this? Presumably you have some use-case in mind?

Reply:
The short summary is if you have a really complex peripheral:

```rust
peripheral! {
    Foo {
        0x0 => ctrl: u32 { Read, Write },
        0x4 => frequency: u16 { Write },
        /* 24 more registers here */
        0x3C => status: u8 { Read },
        /* 12 more registers here */
    }
}
```

and you have a unit test where you just want to fake the `status` register, you still end up needing to specify all the type aliases:

```rust
impl Foo for FakeFoo {
    type ctrl<'s> = FakeRegister<'s, u32, (), Safe, Safe>;
    fn ctrl(&self) -> Self::ctrl<'_> {
        FakeRegister::new()
    }
    type frequency<'s> = FakeRegister<'s, u16, (), NoAccess, Safe>;
    fn frequency(&self) -> Self::frequency<'_> {
        FakeRegister::new()
    }
    /* 24 more registers here */
    type status: u8 = FakeRegister<'s, u8, (), Safe, NoAccess>;
    fn status(&self) -> Self::status<'_> {
        FakeRegister::new().on_read(|_| 0xFF)
    }
    /* 12 more registers here */
}
```

To avoid that, we could have a separate macro that implements all the unneeded fields for you. You might want it to look like:
```rust
#[tock_registers::FillInMissingFieldsForMe]
impl Foo for FakeFoo {
    type status: u8 = FakeRegister<'s, u8, (), Safe, NoAccess>;
    fn status(&self) -> Self::status<'_> {
        FakeRegister::new().on_read(|_| 0xFF)
    }
    
    // The other 38 registers are filled in by the procedural macro.
}
```

However, the procedural macro needs to know what all the fields -- and their types -- are. So instead, you pass in the `macro_rules!` macro:
```rust
#[tock_registers::FillInMissingFieldsForMe(Foo)] // This Foo is the macro_rules! macro
impl Foo for FakeFoo {
    type status: u8 = FakeRegister<'s, u8, (), Safe, NoAccess>;
    fn status(&self) -> Self::status<'_> {
        FakeRegister::new().on_read(|_| 0xFF)
    }
    
    // The other 38 registers are filled in by the procedural macro.
}
```

I'm fairly confident that the above can be implemented. It's just a lot of work, so I'm putting it out of the scope of this PR, and just emitting the `macro_rules!` macro so that adding it in the future isn't a breaking change.

Reply:
Makese sense